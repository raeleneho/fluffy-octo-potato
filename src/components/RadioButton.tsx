import { Box, HStack, Text, useRadio } from "@chakra-ui/react";
import { Role } from "../api/CustomerClient";

interface RadioButtonProps<Type> {
  // label?: string;
  checked: boolean;
  value: Type;
  children?: React.ReactElement | string;
  onChange?: (newValue: Type) => void;
}

export function RadioButton<Type extends string>(props: any) {
  const { getInputProps, getRadioProps } = useRadio(props);
  const input = getInputProps();
  const radio = getRadioProps();

  return (
    <Box as="label" display="block">
      <input {...input} />
      <Box
        {...radio}
        borderRadius="lg"
        p={3}
        _checked={{
          bg: "brandBlue.100",
          ".radio-inner": {
            background: "brandBlue.500",
          },
        }}
        _focus={{
          borderBlockColor: "brandBlue.500",
        }}
      >
        <HStack gap="12px">
          <Box
            width="20px"
            height="20px"
            p="2px"
            border="1px solid grey"
            backgroundColor="white"
            borderRadius="50%"
          >
            <Box
              className="radio-inner"
              width="100%"
              height="100%"
              p="2px"
              borderRadius="50%"
            ></Box>
          </Box>
          <Text textTransform="capitalize">{props.children}</Text>
        </HStack>
      </Box>
    </Box>
  );
}

export default RadioButton;
