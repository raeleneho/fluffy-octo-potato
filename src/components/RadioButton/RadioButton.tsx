import { Box, HStack, Text, UseRadioProps, useRadio } from "@chakra-ui/react";

interface RadioButtonProps extends UseRadioProps {
  children?: React.ReactElement | string;
}

export function RadioButton(props: RadioButtonProps) {
  const { getInputProps, getRadioProps } = useRadio(props);
  const input = getInputProps();
  const radio = getRadioProps();

  return (
    <Box
      as="label"
      display="block"
      data-testid={props.value?.replaceAll(" ", "")}
    >
      <input {...input} />
      <Box
        {...radio}
        borderRadius="lg"
        p={3}
        transition="background 0.3s"
        _checked={{
          bg: "brandBlue.100",
          ".radio-outer": {
            borderColor: "brandBlue.500",
          },
          ".radio-inner": {
            background: "brandBlue.500",
          },
        }}
        _focus={{
          borderBlockColor: "brandBlue.500",
        }}
      >
        <HStack gap="12px">
          <Box
            className="radio-outer"
            width="20px"
            height="20px"
            p="3px"
            border="1px"
            borderColor="gray.200"
            borderRadius="50%"
            transition="border-color 0.3s"
          >
            <Box
              className="radio-inner"
              width="100%"
              height="100%"
              borderRadius="50%"
              transition="background 0.3s"
            ></Box>
          </Box>
          <Text textTransform="capitalize">{props.children}</Text>
        </HStack>
      </Box>
    </Box>
  );
}

export default RadioButton;
