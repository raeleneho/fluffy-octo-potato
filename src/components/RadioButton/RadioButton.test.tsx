import { fireEvent, getByTestId, render } from "@testing-library/react";
import RadioButton from "./RadioButton";
import { useRadioGroup, VStack } from "@chakra-ui/react";
import { useState } from "react";

test("Radio renders", () => {
  const { container } = render(<RadioButton value={"option1"} />);

  expect(container).toBeInTheDocument();

  const radioButton = getByTestId(container, "option1");

  expect(radioButton).toBeVisible();

  expect(radioButton).toMatchSnapshot();
});

test("Change Selection Updates the value", () => {
  const Container = () => {
    const [selectedOption, setSelectedOption] = useState<string>("option 1");

    const handleChange = (value: string) => {
      setSelectedOption(value);
    };

    const options = ["option 1", "option 2"];
    const { getRootProps, getRadioProps } = useRadioGroup({
      name: "radio",
      defaultValue: selectedOption,
      onChange: handleChange,
    });

    const group = getRootProps();

    const radioButtons = (
      <div>
        <VStack {...group} alignItems="stretch">
          {options.map((item) => (
            <RadioButton key={item} {...getRadioProps({ value: item })}>
              {item.toLowerCase()}
            </RadioButton>
          ))}
        </VStack>

        <div data-testid="SelectedOption">{selectedOption}</div>
      </div>
    );

    return radioButtons;
  };
  const { container } = render(<Container />);

  let selectedValue = getByTestId(container, "SelectedOption");
  expect(selectedValue).toHaveTextContent("option 1");

  const uncheckedRadio = getByTestId(container, "option2");
  fireEvent(uncheckedRadio, new MouseEvent("click"));

  selectedValue = getByTestId(container, "SelectedOption");
  expect(selectedValue).toHaveTextContent("option 2");
});
