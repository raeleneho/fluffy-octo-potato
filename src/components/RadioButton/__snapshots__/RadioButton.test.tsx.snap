// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders 1`] = `
<label
  class="css-13o7eu2"
  data-testid="option1"
>
  <input
    id="radio-:r0:"
    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
    type="radio"
    value="option1"
  />
  <div
    aria-hidden="true"
    class="css-mmyk1o"
  >
    <div
      class="chakra-stack css-qlxuh7"
    >
      <div
        class="radio-outer css-12loqvc"
      >
        <div
          class="radio-inner css-a0adc8"
        />
      </div>
      <p
        class="chakra-text css-10iahqc"
      />
    </div>
  </div>
</label>
`;
